pkg/abstractions/common/instance.go

```diff
@@ -287,3 +298,72 @@ func (i *AutoscaledInstance) State() (*AutoscaledInstanceState, error) {
 	state.FailedContainers = failedContainers
 	return &state, nil
 }
+
+func (i *AutoscaledInstance) HandleDeploymentDegraded(stubId string, containers []string) {
+	var state string
+	state, err := i.ContainerRepo.GetStubUnhealthyState(stubId)
+	if err != nil {
+		if err != redis.Nil {
+			log.Printf("<%s> failed to get unhealthy state\n", i.Name)
+			return
+		}
+
+		state = "healthy"
+	}
+
+	if state == "degraded" {
+		return
+	}
+
+	err = i.ContainerRepo.SetStubUnhealthyState(stubId, "degraded")
+	if err != nil {
+		log.Printf("<%s> failed to set unhealthy state\n", i.Name)
+		return
+	}
+
+	go i.EventRepo.PushStubStateDegraded(stubId, state, "failed container retry threshold", containers)
+}
+
+func (i *AutoscaledInstance) HandleDeploymentWarning(stubId string, containers []string) {
+	var state string
+	state, err := i.ContainerRepo.GetStubUnhealthyState(stubId)
+	if err != nil {
+		if err != redis.Nil {
+			log.Printf("<%s> failed to get unhealthy state\n", i.Name)
+			return
+		}
+
+		state = "healthy"
```

Comment:
I'm not sure I would have a new "state", I almost feel like the client can derive the state implicitly, if a certain number of failed containers has been met.